pyogrio/tests/test_geopandas_io.py

```diff
@@ -843,17 +844,56 @@ def test_write_read_null(tmp_path):
     ],
 )
 def test_write_geometry_z_types(tmp_path, wkt, geom_types):
-    from geopandas.array import from_wkt
-
     filename = tmp_path / "test.fgb"
-
     gdf = gp.GeoDataFrame(geometry=from_wkt([wkt]), crs="EPSG:4326")
     for geom_type in geom_types:
         write_dataframe(gdf, filename, geometry_type=geom_type)
         df = read_dataframe(filename)
         assert_geodataframe_equal(df, gdf)
 
 
+@pytest.mark.parametrize("ext", ALL_EXTS)
+@pytest.mark.parametrize(
+    "wkt, exp_geometry_type",
+    [
+        ("Point Z (0 0 0)", "2.5D Point"),
+        ("LineString Z (0 0 0, 1 1 0)", "2.5D LineString"),
+        ("Polygon Z ((0 0 0, 0 1 0, 1 1 0, 0 0 0))", "2.5D Polygon"),
+        ("MultiPoint Z (0 0 0, 1 1 0)", "2.5D MultiPoint"),
+        (
+            "MultiLineString Z ((0 0 0, 1 1 0), (2 2 2, 3 3 2))",
+            "2.5D MultiLineString",
+        ),
+        (
+            "MultiPolygon Z (((0 0 0, 0 1 0, 1 1 0, 0 0 0)), ((1 1 1, 1 2 1, 2 2 1, 1 1 1)))",  # NOQA
+            "2.5D MultiPolygon",
+        ),
+        ("GeometryCollection Z (Point Z (0 0 0))", "2.5D GeometryCollection"),
+    ],
+)
+def test_write_geometry_z_types_auto(tmp_path, ext, wkt, exp_geometry_type):
+    # Shapefile has some different behaviour that other file types
+    if ext == ".shp":
+        if exp_geometry_type == "2.5D GeometryCollection":
+            pytest.skip(f"ext {ext} doesn't support {exp_geometry_type}")
+        elif exp_geometry_type == "2.5D MultiLineString":
+            exp_geometry_type = "2.5D LineString"
+        elif exp_geometry_type == "2.5D MultiPolygon":
+            exp_geometry_type = "2.5D Polygon"
```

Comment:
But so it still writes correctly? (and reads as 3D multi geometries?) 
If so, this might also be a bug in GDAL?

Reply:
Yes, the data is saved correctly. Shapefile doesn't make a distinction between multi and single at file/layer metadata level... so gdal always report single. This is why it is also not needed to promote to multi to save mixed single and multi geometries in shapefile.